src/gt4py/analysis/passes.py

```diff
@@ -406,41 +405,58 @@ def __init__(self):
+            stmt_info.inputs,
+            stmt_info.outputs,
+        )
+        new_ij_block = IJBlockInfo(
+            transform_data.id_generator.new,
+            {interval},
+            [new_interval_block],
+            {**new_interval_block.inputs},
+            set(new_interval_block.outputs),
+            compute_extent=zero_extent,
+        )
+        return DomainBlockInfo(
+            transform_data.id_generator.new,
+            block.iteration_order,
+            set(new_ij_block.intervals),
+            [new_ij_block],
+            {**new_ij_block.inputs},
+            set(new_ij_block.outputs),
+        )
+
+    def _iter_ij_blocks(self, transform_data):
         for block in transform_data.blocks:
-            # Put every statement in a single stage
             for ij_block in block.ij_blocks:
-                for interval_block in ij_block.interval_blocks:
-                    for stmt_info in interval_block.stmts:
-                        interval = interval_block.interval
-                        new_interval_block = IntervalBlockInfo(
-                            transform_data.id_generator.new,
-                            interval,
-                            [stmt_info],
-                            stmt_info.inputs,
-                            stmt_info.outputs,
-                        )
-                        new_ij_block = IJBlockInfo(
-                            transform_data.id_generator.new,
-                            {interval},
-                            [new_interval_block],
-                            {**new_interval_block.inputs},
-                            set(new_interval_block.outputs),
-                            compute_extent=zero_extent,
-                        )
-                        new_block = DomainBlockInfo(
-                            transform_data.id_generator.new,
-                            block.iteration_order,
-                            set(new_ij_block.intervals),
-                            [new_ij_block],
-                            {**new_ij_block.inputs},
-                            set(new_ij_block.outputs),
-                        )
-                        blocks.append(new_block)
+                yield block, ij_block
+
+    def _iter_interval_blocks(self, transform_data):
+        for block, ij_block in self._iter_ij_blocks(transform_data):
+            for interval_block in ij_block.interval_blocks:
+                yield block, interval_block
 
-        transform_data.blocks = blocks
+    def _iter_statement_args(self, transform_data):
+        for block, interval_block in self._iter_interval_blocks(transform_data):
+            for statement in interval_block.stmts:
+                yield transform_data, block, interval_block, statement
+
+    def apply(self, transform_data: TransformData):
+        transform_data.blocks = [
+            self._block_from_statement(*args) for args in self._iter_statement_args(transform_data)
+        ]
```

Comment:
This encapsulation is way to exessive to me. Sure, each part is small and compact, but I have to follow 3 functions with a total of 6 loops to understand a 4 dimensional loop and then the they don't even extract the necessary parts for the `_block_from_statement` function, but just pass the iteration variables on. I agree that the previous version can be improved, but I suggest you use just one function containing the 4 dimensional loop, extract the necessary bits actually _used_ in the `_block_from_statement` (e.g. `ndims` not `transform_data`) there and then pass (via yield) that to `_block_from_statement` explicitly here.

Reply:
As mentioned in the PR description, I used `flake8-cognitive-complexity` as a measure here. It gave a very high complexity score to the quadruple for loop. I will investigate why.

Reply:
In the meantime, I replaced it with a visitor-like class, because that matches what we are doing far better than nested for loops and increases readability by sticking to a pattern that comes up frequently in this code base anyway.

Reply:
For cognitive complexity: https://blog.sonarsource.com/cognitive-complexity-because-testability-understandability explains it. Each control flow breaking statement increments according to it's nesting level:

```python
for ...:  # +1
    for ...:  # +2 = 3
        for ...:  # +3 = 6
            for ...:  # +4 = 10
```

So a quadruple-nested for loop is always past the default threshold of >=7, above which `flake8-cognitive-complexity` warns. I do agree by the way with that understandability goes down the drain the quicker the deeper the nesting.